### ISSUE TITLE
Multiple rabbit MQ reports all connections are unreachable if one connection is down

### ISSUE THREAD

**Issue by chevillanos** [Xabaril/AspNetCore.Diagnostics.HealthChecks#764]
Please, fill the following sections to help us fix the issue

**What happened**: Multiple rabbit MQ reports all are unreachable if one connection is down.
Usage: 
```
services.AddHealthChecks()
.AddRabbitMQ(Configuration["rabbitconnStr1"].ToString(), null, "rabbitMqNode1", timeout: TimeSpan.FromSeconds(15))
.AddRabbitMQ(Configuration["rabbitconnStr2"].ToString(), null, "rabbitMqNode2", timeout: TimeSpan.FromSeconds(15));
```

**What you expected to happen**: I am expecting rabbitMqNode1 is healthy if rabbitMqNode2 is down.

**How to reproduce it (as minimally and precisely as possible)**: The usage code above and intentionally stopping the RabbitMq found in rabbitconnStr2 connection while rabbitconnStr1 is up and accessible.
Tested this on a different project while rabbitconnStr1 is reporting as healthy without rabbitconnStr2 declared on the same scenario. Am I chaining them correctly as stated in the readme documentation?

I've also tried having a separate connection with services separately but faces similar behavior:

```
.AddSingleton<IConnectionFactory>(sp=>
            new ConnectionFactory()
            {
                Uri = new Uri(Configuration["rabbitconnStr1"].ToString()),
                AutomaticRecoveryEnabled = true
            })
        .AddHealthChecks()
        .AddRabbitMQ(name: "rabbitMqNode1");

.AddSingleton<IConnectionFactory>(sp=>
            new ConnectionFactory()
            {
                Uri = new Uri(Configuration["rabbitconnStr2"].ToString()),
                AutomaticRecoveryEnabled = true
            })
        .AddHealthChecks()
        .AddRabbitMQ(name: "rabbitMqNode2");
```

*Edit: If I declare rabbitMqNode2 first and rabbitMqNode1 at the bottom shows both endpoints are healthy. I assume whichever comes last overrides the initial declaration? If this is the case, how can I monitor two or more Uri's at the same time?

**Environment**: dev / local

- .NET Core version net 5.0
- Healthchecks version 5.0.1
- Operative system: Windows 10
- Others: Visual Studio 2019 16.9.2

**Comment by georgy93** [comment]
Hello!
I just hit the same problem yesterday and I think the explanation is that the RabbitMQHealthCheck is retrieved by the ASP NET Core DI framework, which resolves the last registered type for a given Interface/Type.

Looking at the HealthCheck registering code for RabbitMQ we can see that we register RabbitMQHealthCheck type in the service container
```
builder.Services.AddSingleton(sp => new RabbitMQHealthCheck(.....);
```

And then retrieve an instance of the type RabbitMQHealthCheck `  sp => sp.GetRequiredService<RabbitMQHealthCheck>(),`

I believe the author of the package decided to register the RabbitMQHealthCheck in the DI and as a Singleton, because creation of a new RabbitMQHealthCheck everytime will lead to opening a new IConnection and this is rather expensive operation and uses more resources on the Rabbit server. Opening IChannel `_connection.CreateModel()` is way cheaper.


> I've also tried having a separate connection with services separately but faces similar behavior:

> AddSingleton<IConnectionFactory>(sp=>
            new ConnectionFactory()
            {
                Uri = new Uri(Configuration["rabbitconnStr1"].ToString()),
                AutomaticRecoveryEnabled = true
            })
        .AddHealthChecks()
        .AddRabbitMQ(name: "rabbitMqNode1");

> .AddSingleton<IConnectionFactory>(sp=>
            new ConnectionFactory()
            {
                Uri = new Uri(Configuration["rabbitconnStr2"].ToString()),
                AutomaticRecoveryEnabled = true
            })
        .AddHealthChecks()
        .AddRabbitMQ(name: "rabbitMqNode2");

This will still not work for the same reason. Even if a new RabbitMQHealthCheck instance was created fresh everytime, it still depends on a single IConnection / IConnectionFactory constructor parameter, and the last registered type of it is resolved.

**Comment by chevillanos** [comment]
I moved away from the AddRabbitMQ and just used the HealthCheckRegistration instead and passed on a ConnectionFactory for each node
thanks georgy

**Comment by DPDmancul** [comment]
A solution to allow health checking multiple rabbit connections without instantiating a connection each time could be the following:

```c#
    public static IHealthChecksBuilder AddRabbitMQ(
        this IHealthChecksBuilder builder,
        Uri rabbitConnectionString,
        SslOption? sslOption = default,
        string? name = default,
        HealthStatus? failureStatus = default,
        IEnumerable<string>? tags = default,
        TimeSpan? timeout = default)
    {
        var rabbitMQHalthCheck = new(rabbitConnectionString, sslOption);
        return builder.Add(new HealthCheckRegistration(
            name ?? NAME,
            _ => rabbitMQHalthCheck,
            failureStatus,
            tags,
            timeout));
```

This way the connection is established only once, at the addition of the health check, but multiple (different) connections are allowed. Unfortunately this instantiates the connection even if the health check route is never called. To solve this we can use `Lazy`:

```c#
    public static IHealthChecksBuilder AddRabbitMQ(
        this IHealthChecksBuilder builder,
        Uri rabbitConnectionString,
        SslOption? sslOption = default,
        string? name = default,
        HealthStatus? failureStatus = default,
        IEnumerable<string>? tags = default,
        TimeSpan? timeout = default)
    {
        var rabbitMQHalthCheck = new Lazy<RabbitMQHealthCheck>(() => new(rabbitConnectionString, sslOption));
        return builder.Add(new HealthCheckRegistration(
            name ?? NAME,
            _ => rabbitMQHalthCheck.Value,
            failureStatus,
            tags,
            timeout));
    }
```

Unfortunately `Lazy` does not allow passing parameters, and so to capture the service provider we have to manually reimplement lazy:

```c#
    public static IHealthChecksBuilder AddRabbitMQ(
        this IHealthChecksBuilder builder,
        Func<IServiceProvider, IConnection> connectionFactory,
        string? name = default,
        HealthStatus? failureStatus = default,
        IEnumerable<string>? tags = default,
        TimeSpan? timeout = default)
    {
        object rabbitMQHealthCheckLock = new();
        RabbitMQHealthCheck? rabbitMQHealthCheck = null;
        return builder.Add(new HealthCheckRegistration(
            name ?? NAME,
            sp =>
            {
                if (rabbitMQHealthCheck is null)
                {
                    lock (rabbitMQHealthCheckLock)
                    {
                        if (rabbitMQHealthCheck is null)
                        {
                            rabbitMQHealthCheck = new RabbitMQHealthCheck(connectionFactory(sp));
                        }
                    }
                }
                return rabbitMQHealthCheck;
            },
            failureStatus,
            tags,
            timeout));
    }
```

If this is fine, I can made a PR

**Comment by sungam3r** [comment]
This is not a question of RabbitMQ (or some other) health check. This is a question about general design - should or not health check allow to work with different servers/databases/services/etc. Note that many health checks from that repo do not support such cases. RabbitMQ HC (particularly extensions methods part of it) is just one part of the whole picture. You may find other places by typing `AddSingleton` in the search bar. I'm not sure that in all cases you can achieve the desired behavior. In some cases, this will lead either to the complication of the code or to the loss of performance (maybe both).

**Comment by DPDmancul** [comment]
Yes I know. But some checks (e.g. NpgSql) are already migrating away from Singleton (https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/commit/89dcf77b5e06497c599a7a0cc75ed70937269b1d#diff-265d89a04891e22c10bb72318343f9df8cb007174a68c40edc94c1f6ed9ddfd8L69). In NpgSql, Redis (https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/eeb24329d33fb0332ca22ac916516348bd646205/src/HealthChecks.Redis/DependencyInjection/RedisHealthCheckBuilderExtensions.cs#L62), Azure (https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/eeb24329d33fb0332ca22ac916516348bd646205/src/HealthChecks.AzureStorage/DependencyInjection/AzureStorageHealthCheckBuilderExtensions.cs#L47), and many others a new connection is instantiated each time the health check is called. They so allow multiple connections. To uniform RabbitMQ check would not be possible because of RabbitMQ being not happy with frequent connection/disconnection, so I suggested the Lazy way to obtain the same benefits.
This would not break the previous behavior and do not alter the general design, since there is already no consistency: some checks use singletons, others don't. Not using singleton covers more use cases, so I suggested this path.

**Comment by sungam3r** [comment]
> and many others a new connection is instantiated each time the health check is called.

Yes. I would say this is one of two primary patterns - either instantiate "fresh" objects required to HC each time or cache them (singleton registration effectively does it).

> Not using singleton covers more use cases

Yes but in practice 99% of apps do not need to connect to multiple instances of RabbitMQ or whatever external service. `RedisHealthCheck` from your example acts as thin wrapper around long-lived cached dictionary of `ConnectionMultiplexer`s.


I want to say that if we start looking for the proposed path, then we need to understand in advance how to do this for every (or at least for most) health checks. I'm fine to go this way to allow multiple configurations of the same health check but not at the cost of performance loss. PR is welcome and I suggest to open new issue and not continue dicsussion here.
